src/databricks/labs/dqx/llm/llm_core.py

```diff
@@ -0,0 +1,304 @@
+        Get rules as a JSON string.
+
+        Args:
+            schema_info (str): JSON string containing table schema with column names, types, and sample data.
+            business_description (str): Natural language description of data quality requirements.
+            available_functions (str): JSON string of available DQX check functions.
+
+        Returns:
+            str: A JSON string representing the generated data quality rules.
+        """
+        result = self.forward(schema_info, business_description, available_functions)
+        return result.quality_rules
+
+
+def _configure_dspy_model(model: str, api_key: str = "", api_base: str = ""):
+    """
+    Configure the Dspy language model.
+
+    Args:
+        model (str): The model to use for the Dspy language model.
+        api_key (str): The API key for the model. Not required by Databricks foundational models.
+        api_base (str): The API base URL for the model. Not required by Databricks foundational models.
+    """
+    language_model = dspy.LM(
+        model=model,
+        model_type="chat",
+        api_key=api_key,
+        api_base=api_base,
+        max_retries=3,
+    )
+    dspy.configure(lm=language_model)
+
+
+def validate_generated_rules(expected: str, actual: str) -> float:
+    """
+    Validate generated rules with granular scoring for better optimizer feedback.
+
+    Scoring breakdown:
+        - JSON parsing (20%): Checks if the actual output can be parsed as valid JSON.
+        - Rules validation (50%): Ensures the rules pass DQX validation checks.
+        - Content similarity (30%): Compares the actual rules content with the expected rules.
+
+    Args:
+        expected (str): JSON string of the expected rules.
+        actual (str): JSON string of the actual generated rules.
+
+    Returns:
+        float: A score between 0.0 and 1.0 representing the quality of the generated rules.
+    """
+    total_score = 0.0
+
+    # Score weights
+    json_weight = 0.2
+    rule_weight = 0.5
+    similarity_weight = 0.3
+
+    expected_rules = json.loads(expected)
+
+    # Json parsing score (20%)
+    try:
+        actual_rules = json.loads(actual)
+        total_score += json_weight
+        logger.info(f"✓ JSON parsing successful (+{json_weight:.1f})")
+    except json.JSONDecodeError as e:
+        logger.warning(f"✗ JSON parsing failed: {e}")
+        logger.debug(f"  Raw output: {repr(actual[:200])}")
+        # Early return if we can't parse JSON at all
+        return total_score
+
+    # Rules validation score (50%)
+    validation_status = DQEngineCore.validate_checks(actual_rules)
+    if not validation_status.has_errors:
+        total_score += rule_weight
+        logger.info(f"✓ Rules validation passed (+{rule_weight:.1f})")
+    else:
+        logger.warning(f"✗ Rules validation errors: {validation_status.errors}")
+
+    # Content similarity score (30%)
+    similarity_score = _calculate_rule_similarity(expected_rules, actual_rules)
```

Comment:
this is not needed: `DQEngineCore.validate_checks(actual_rules)` validates the generated rule is correct. There should be no need to compare against expected. It would also be problematic to maintain because a lot of arguments can be optional and providing examples of all possible combinations won't be possible